buildreport/buildreport.go

```diff
@@ -0,0 +1,454 @@
+	return githubutil.Retry(func() error {
+		edit, _, err := client.Issues.Edit(ctx, owner, repoName, issue, &github.IssueRequest{Body: &body})
+		if err != nil {
+			return err
+		}
+		log.Printf("Edit successful:\n%v\n", edit.ID)
+		return nil
+	})
+}
+
+// Notify determines if a notification is necessary for the given status update and sends it.
+func Notify(ctx context.Context, owner string, repoName string, pat string, issue int, s State) error {
+	var notification string
+	switch s.Status {
+	case SymbolFailed:
+		notification = "Build failed!\n"
+	case SymbolSucceeded:
+		switch s.Name {
+		case releaseBuildPipelineName:
+			notification = "Completed releasing one version! If every version's release-build has finished, approve the release-go-images build to continue.\n"
+		case releaseImagesPipelineName:
+			notification = "Completed building all images! " +
+				"Before you [announce](https://github.com/microsoft/go-infra/blob/main/docs/release-process/instructions.md#making-the-internal-announcement), " +
+				"confirm the MAR/MCR images are updated using commands like these:\n\n" +
+				"```\n" +
+				"image=mcr.microsoft.com/oss/go/microsoft/golang:1-bullseye\n" +
+				"docker pull $image\n" +
+				"docker run -it --rm $image go version\n" +
+				"```\n"
+		}
+	}
+
+	if notification != "" {
+		client, err := githubutil.NewClient(ctx, pat)
+		if err != nil {
+			return err
+		}
+
+		if err := githubutil.Retry(func() error {
+			c := commentBody{reports: []State{*&s}}
+			body, err := c.body()
+			if err != nil {
+				return err
+			}
+
+			notification += body +
+				"\n<sub>See the [issue description](" +
+				"https://github.com/" + owner + "/" + repoName + "/issues/" + strconv.Itoa(issue) +
+				") for the latest build status.</sub>"
+
+			notificationComment, _, err := client.Issues.CreateComment(
+				ctx, owner, repoName, issue, &github.IssueComment{Body: &notification})
+			if err != nil {
+				return err
+			}
+			log.Printf("Notification comment: %v\n", *notificationComment.HTMLURL)
+			return nil
+		}); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// State is the status of one entry in the report.
+type State struct {
+	// ID of the report. If an AzDO build, the AzDO Build ID.
+	ID string
+	// Version this report is associated with.
+	Version string
+	// Name of this type of report. If an AzDO build, the pipeline name.
+	Name string
+	// URL that the ID should link to. If an AzDO build, the main build page.
+	URL string
+	// Status represents the status.
+	Status string
+
+	LastUpdate time.Time
+	StartTime  time.Time
+}
+
+func (s *State) updateFrom(source State) {
+	if source.Version != "" {
+		s.Version = source.Version
+	}
+	if source.Name != "" {
+		s.Name = source.Name
+	}
+	if source.URL != "" {
+		s.URL = source.URL
+	}
+	if source.Status != "" {
+		s.Status = source.Status
+	}
+	if source.LastUpdate != (time.Time{}) {
```

Comment:
The idiomatic way to do this check is using [time.IsZero()](https://pkg.go.dev/time#Time.IsZero).

Reply:
Thanks, definitely looks nicer.

I mentioned in standup that doing this change made it behave differently, but it's just because I thoughtlessly slapped in `if source.LastUpdate.IsZero()` rather than `if !source.LastUpdate.IsZero()`. 😄 Fixed.

`IsZero` does seem to be a little different because it compares actual time rather than directly comparing the `location` pointers, but this is a good thing for our sanity, and it doesn't change any behavior I care about.